athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/storage/StorageMetadata.java

```diff
@@ -0,0 +1,287 @@
+     */
+    private List<Field> getFields(String bucketName, String prefix, String format, BufferAllocator allocator)
+    {
+        LOGGER.info("Getting table fields for object {}.{}", bucketName, prefix);
+        Optional<String> file = getStorageFiles(bucketName, prefix);
+        if (file.isPresent()) {
+            return getFileSchema(bucketName, file.get(), FileFormat.valueOf(format.toUpperCase()), allocator).getFields();
+        }
+
+        throw new IllegalArgumentException("No object found for the table name '" + prefix + "' under bucket " + bucketName);
+    }
+
+    /**
+     * Retrieves a list of StorageSplit that essentially contain the list of all files for a given table type in a storage location
+     *
+     * @param locationUri location uri
+     * @return A list of files
+     */
+    public List<String> getStorageSplits(URI locationUri)
+    {
+        List<String> fileList = new ArrayList<>();
+        String bucketName = locationUri.getAuthority();
+        String path = locationUri.getPath().startsWith("/") ? locationUri.getPath().substring(1) : locationUri.getPath();
+        Page<Blob> blobs = storage.list(bucketName, prefix(path));
+        for (Blob blob : blobs.iterateAll()) {
+            if (isBlobFile(blob)) {
+                fileList.add(bucketName + "/" + blob.getName());
+            }
+        }
+        return fileList;
+    }
+
+    /**
+     * Retrieves a list of partition folders from the GCS bucket based on partition.pattern Table parameter and partition keys set forth in Glue table. If the summary from the
+     * constraints is empty (no where clauses or unsupported clauses), it will essentially return all the partition folders from the GCS bucket. If there is any constraints to
+     * apply, it will apply constraints to filter selected partition folder, to narrow down the data load
+     *
+     * @param schema An instance of {@link Schema} that describes underlying Table's schema
+     * @param tableInfo Name of the table
+     * @param constraints An instance of {@link Constraints}, captured from where clauses
+     * @param awsGlue An instance of {@link AWSGlue}
+     * @return A list of {@link Map<String, String>} instances
+     * @throws URISyntaxException Throws if any occurs during parsing Uri
+     */
+    public List<Map<String, String>> getPartitionFolders(Schema schema, TableName tableInfo, Constraints constraints, AWSGlue awsGlue)
+            throws URISyntaxException
+    {
+        LOGGER.info("Getting partition folder(s) for table {}.{}", tableInfo.getSchemaName(), tableInfo.getTableName());
+        Table table = GcsUtil.getGlueTable(tableInfo, awsGlue);
+        List<Field> partitionFields = schema.getFields().stream()
+                .filter(field -> table.getPartitionKeys().stream()
+                        .map(Column::getName)
+                        .collect(Collectors.toList()).contains(field.getName()))
+                .collect(Collectors.toList());
```

Comment:
Is this filter that you added in the most recent update here necessary?

I'm seeing that `FilterExpressionBuilder::getExpressions()` will basically end up doing the same thing again here:
```
 62                 ValueSet valueSet = constraints.getSummary().get(column.getName());
```

Additionally, if it really is necessary, then this should be rewritten to avoid looping through `table.getPartitionKeys()` for every element of `schema.getFields()`, like this:

```
        var partitionFieldNames = table.getPartitionKeys().stream().map(Column::getName).collect(Collectors.toSet());
        List<Field> partitionFields = schema.getFields().stream()
            .filter(field -> partitionFieldNames.contains(field.getName()))
            .collect(Collectors.toList());
```

Reply:
Here I made the change for you guys:
https://github.com/trianz-azam/aws-athena-query-federation/commit/03623fb6d59c33491c5b6f80dab98688dec57d13

However, please verify whether or not the filtering is even necessary to begin with, if not, we can just get rid of the filtering code here.

Reply:
This filtering is required, as we are dealing with only partition column constraints in this flow. It will be unnecessary to do all other processes for all columns.

Reply:
Ok, so that means that `constraints` contains constraints for more than just partitioned columns and we are trying to just ignore all constraints other than the ones for partitioned columns, right?

So that means `FilterExpressionBuilder` can be altered to just take in the partitioned columns directly (`table.getPartitionKeys()`) rather than just taking in a `List<Field>`, since we don't use the `Field` type for anything other than just the name anyway and also we aren't using the `partitionedField` that we're creating other than to pass into `FilterExpressionBuilder` anyway, meaning we can eliminate a bunch of redundant iterations.

I will update the code to show you what I mean.

Reply:
Just made the change: https://github.com/trianz-azam/aws-athena-query-federation/commit/c7d9531c74addfa6c237cea2e4c24f61f7d73dce

Please test all of the latest changes on your branch and make sure that things still work.

Reply:
Thanks, It is working fine.